src/components/document/document-workspace.tsx

```diff
@@ -135,38 +162,80 @@ export class DocumentWorkspaceComponent extends BaseComponent<IProps, {}> {
       >
         Click or drag an item in the right tabs to show it here
       </div>
     );
   }
 
-  private handleDragOver = (e: React.DragEvent<HTMLDivElement>) => {
-    if (e.dataTransfer.types.find((type) => type === DocumentDragKey)) {
+  private isAcceptableImageDrag = (e: React.DragEvent<HTMLDivElement>) => {
+    // make sure we have a primary document to drop onto
+    return !!this.getPrimaryDocument(this.stores.ui.problemWorkspace.primaryDocumentKey);
+  }
+
+  private handleDragOverWorkspace = (e: React.DragEvent<HTMLDivElement>) => {
+    this.imageDragDrop.dragOver(e);
+  }
+
+  private handleDragOverSide = (e: React.DragEvent<HTMLDivElement>) => {
+    if (this.imageDragDrop.dragOver(e) || e.dataTransfer.types.find((type) => type === DocumentDragKey)) {
       e.preventDefault();
     }
   }
 
-  private handleDrop = (side: WorkspaceSide) => {
+  private handleDropSide = (side: WorkspaceSide) => {
     return (e: React.DragEvent<HTMLDivElement>) => {
       const {ui, documents} = this.stores;
-      const {problemWorkspace} = ui;
       const documentKey = e.dataTransfer && e.dataTransfer.getData(DocumentDragKey);
-      const document = documentKey ? documents.getDocument(documentKey) : null;
-      if (document) {
-        if (side === "primary") {
-          problemWorkspace.setPrimaryDocument(document);
+      if (documentKey) {
+        const {problemWorkspace} = ui;
+        const document = documents.getDocument(documentKey);
+        if (document) {
+          if (side === "primary") {
+            problemWorkspace.setPrimaryDocument(document);
+          }
+          else {
+            problemWorkspace.setComparisonDocument(document);
+          }
         }
-        else {
-          problemWorkspace.setComparisonDocument(document);
+      }
+      else {
+        // try to get the row it was dropped on
+        let rowNode = e.target as HTMLElement | null;
+        while (rowNode && (rowNode.className !== "tile-row")) {
+          rowNode = rowNode.parentNode as HTMLElement | null;
         }
+        const rowId = (rowNode && rowNode.dataset && rowNode.dataset.rowId) || undefined;
+        this.handleImageDrop(e, rowId);
       }
     };
   }
 
+  private handleImageDrop = (e: React.DragEvent<HTMLDivElement>, rowId?: string) => {
+    const {ui} = this.stores;
+    this.imageDragDrop.drop(e)
+      .then((dropUrl) => {
+        const {problemWorkspace} = ui;
+        const primaryDocument = this.getPrimaryDocument(problemWorkspace.primaryDocumentKey);
+        if (primaryDocument) {
+          const rowIndex = rowId ? primaryDocument.content.getRowIndex(rowId) : undefined;
+          primaryDocument.content.addTile("image", {
+            imageUrl: dropUrl,
+            insertRowInfo: {
+              // insert the tile after the row it was dropped on otherwise add to end of document
+              rowInsertIndex: (rowIndex ? rowIndex + 1 : primaryDocument.content.rowOrder.length)
```

Comment:
```suggestion
              rowInsertIndex: (rowIndex != null ? rowIndex + 1 : primaryDocument.content.rowOrder.length)
```
I think we still need the `!= null` to handle the `rowIndex === 0` case properly unless I'm missing something.